### ISSUE TITLE
Move Python-specific interface code from main library to Python interface repo

### ISSUE THREAD

**Issue by imciner2** [osqp/osqp-python#76]
The merging of https://github.com/osqp/osqp/pull/396 in the main OSQP library repo will allow us to move the Python-specific printing functions to the Python repo, removing the need to include the Python headers and code in the main library.

This is also possible to do with the memory allocation functions currently, and hopefully the timing/interrupt functions will move to this as well, so we can then move all Python-interface specific code to the interface and remove it from the main library repo.

**Comment by imciner2** [comment]
The Python-specific code for the printing functions is:
```
# include <Python.h>
# define c_print(...)                              \
  {                                                  \
    PyGILState_STATE gilstate = PyGILState_Ensure(); \
    PySys_WriteStdout(__VA_ARGS__);                  \
    PyGILState_Release(gilstate);                    \
  }
```

This should be placed into a header file, and then the OSQP_CUSTOM_PRINTING CMake option should be set to the header file containing that code.

**Comment by imciner2** [comment]
The Python-specific memory management code is
```
// Define memory allocation for python. Note that in Python 2 memory manager
// Calloc is not implemented
#   include <Python.h>
#   if PY_MAJOR_VERSION >= 3
// https://docs.python.org/3/c-api/memory.html
// The following function sets are wrappers to the system allocator. These functions are thread-safe, the GIL does not need to be held.
// The default raw memory allocator uses the following functions: malloc(), calloc(), realloc() and free(); call malloc(1) (or calloc(1, 1)) when requesting zero bytes.
#    define c_malloc PyMem_RawMalloc
#    define c_calloc PyMem_RawCalloc
#    define c_free PyMem_RawFree
#    define c_realloc PyMem_RawRealloc
#   else  /* if PY_MAJOR_VERSION >= 3 */
#   define c_malloc PyMem_Malloc
#   define c_free PyMem_Free
#   define c_realloc PyMem_Realloc
static void* c_calloc(size_t num, size_t size) {
    void *m = PyMem_Malloc(num * size);
    memset(m, 0, num * size);
    return m;
}
#   endif /* if PY_MAJOR_VERSION >= 3 */
```

Which should be included in a header file, and then the `OSQP_CUSTOM_MEMORY` CMake variable should be set with that header file's name.

**Comment by vineetbansal** [comment]
Implemented in commit fdb65e9a0126a2352f34ab6ea86bf7eacb8ab08b
